lib/Jsrt/ChakraCore.h

```diff
@@ -25,4 +25,142 @@
+/// holds the exception. otherwise the referencingModule is ready and the host should schedule execution afterwards.
+/// referencingModule       : the referencing module that have finished running ModuleDeclarationInstantiation step.
+/// </remarks>
+/// <param name="exceptionVar">If nullptr, the module is successfully initialized and host should queue the execution job
+///                           otherwise it's the exception object.</param>
+/// <returns>
+///     true if the operation succeeded, false otherwise.
+/// </returns>
+typedef JsErrorCode(CHAKRA_CALLBACK * NotifyModuleReadyCallback)(_In_opt_ JsModuleRecord referencingModule, _In_opt_ JsValueRef exceptionVar);
+
+/// <summary>
+///     Initialize a ModuleRecord from host
+/// </summary>
+/// <remarks>
+///     Bootstrap the module loading process by creating a new module record.
+/// </remarks>
+/// <param name="referencingModule">The referencingModule as in HostResolveImportedModule (15.2.1.17). nullptr if this is the top level module.</param>
+/// <param name="normalizedSpecifier">The host normalized specifier. This is the key to a unique ModuleRecord.</param>
+/// <param name="specifierLength">The length of normalizedSpecifier.</param>
+/// <param name="moduleRecord">The new ModuleRecord created. The host should not try to call this API twice with the same normalizedSpecifier.
+///                           chakra will return an existing ModuleRecord if the specifier was passed in before.</param>
+/// <returns>
+///     The code <c>JsNoError</c> if the operation succeeded, a failure code otherwise.
+/// </returns>
+CHAKRA_API
+JsInitializeModuleRecord(
+    _In_opt_ JsModuleRecord referencingModule,
+    _In_ JsValueRef normalizedSpecifier,
+    _Outptr_result_maybenull_ JsModuleRecord* moduleRecord);
+
+/// <summary>
+///     Parse the module source
+/// </summary>
+/// <remarks>
+/// This is basically ParseModule operation. It is slightly different in that the ModuleRecord was initialized earlier, and passed in as an argument. 
+/// </remarks>
+/// <param name="requestModule">The ModuleRecord that holds the parse tree of the source code.</param>
+/// <param name="sourceContext">A cookie identifying the script that can be used by debuggable script contexts</param>
+/// <param name="script">The source script to be parsed, but not executed in this code.</param>
+/// <param name="scriptLength">The source length of sourceText. The input might contain nested null.</param>
+/// <param name="sourceFlag">The type of the source code passed in. It could be UNICODE or utf8 at this time.</param>
+/// <param name="exceptionValueRef">The error object if there is parse error.</param>
+/// <returns>
+///     The code <c>JsNoError</c> if the operation succeeded, a failure code otherwise.
+/// </returns>
+CHAKRA_API
+JsParseModuleSource(
+    _In_ JsModuleRecord requestModule,
+    _In_ JsSourceContext sourceContext,
```

Comment:
We need to differentiate nested module vs. root module. I'm currently letting the host to maintain the dictionary of specifier to modulerecord mapping but I might want to keep a record in chakracore to make consistency check as we..